Add Google Summer of Code page (#769)
mattermost/mattermost-developer-documentation: site/content/contribute/gsoc/_index.md

    -   Alejandro García Montoro ([@alejandro.garcia](https://community.mattermost.com/core/messages/@alejandro.garcia))
    -   Martin Kraft ([@martin.kraft](https://community.mattermost.com/core/messages/@martin.kraft))
-   A product manager that will support the product decisions across all projects:
    -   Eric Sethna ([@eric.sethna](https://community.mattermost.com/core/messages/@eric.sethna))
-   The organization administrators:
    -   Emily Cook ([@emily.cook](https://community.mattermost.com/core/messages/@emily.cook))
    -   Alejandro García Montoro ([@alejandro.garcia](https://community.mattermost.com/core/messages/@alejandro.garcia))

If you are [eligible to participate as a student](https://summerofcode.withgoogle.com/get-started/), take a look at the list of ideas below, and make sure to check if Mattermost is among the selected organizations on March 9!
To keep track of the latest news from our side, make sure to join the [Google Summer of Code 2021 channel](https://community-daily.mattermost.com/core/channels/google-summer-of-code-2021), where we are centralizing all the communication related to the program.

> If we are finally selected as an organization, you will need to send your student proposals to [gsoc@mattermost.com](mailto:gsoc@mattermost.com). We will hopefully share more details in the coming weeks!

Reviewer: ```suggestion
If we are selected as a participating organization in the program, the next step includes sending us your project proposal to us at [gsoc@mattermost.com](mailto:gsoc@mattermost.com). We're looking forward to  sharing more details with you in the coming weeks!
```

Here’s a sneak peak of what’s coming next:

-   March 9: Organizations announced
-   March 29 - April 13: Student application period
-   April 13 - May 17: Application review period
-   May 17: Student projects announced

> -   May 17 - June 7: Community bonding

Reviewer: Not entirely sure what "community bonding" means or implies here.

Author: It's how the GSoC site calls that period: it's the period where the student knows they're selected but before the coding period begins.

> -   June 7 - August 16: Coding

Reviewer: ```suggestion
-   June 7 - August 16: Ready? Set? CODE!
```

Author: I like the energy here, but I was trying to mimic the period names used by GSoC

Reviewer: Entirely understandable :)

> Make sure to take a look at [the full timeline](https://summerofcode.withgoogle.com/how-it-works/#timeline) for more details!

Reviewer: ```suggestion
Make sure to take a look at [the full Google Summer of Code timeline](https://summerofcode.withgoogle.com/how-it-works/#timeline) for more details!
```

# Ideas list

## Core features

All ideas in the _Core feature_ category propose changes directly to the product, either in the server or the web client.

### TypeScript definitions from Go code

Declaring the same data model in both sides of a web application is an old problem. In our case, the data structures in the server are declared in Go, while those same models are defined in the client side in Typescript. Although the relation is not always one to one, there are lots of places where we could leverage an automatic creation of Typescript types from a struct in Go, and having such a tool would make the life of all our contributors much easier.

The main idea in this project is to investigate the existing solutions to this problem, try to find the best tool that suits the needs of the project and implement it. A nice to have feature in this tool would be a way of not only defining the types, but of keeping them in sync.

> -   **Expected outcomes**: Build tool to generate Typescript types from Go structs.

Author: Oh, here I meant a build tool, as in "a tool to use during the plugin building process"

> -   **Expected outcomes**: Build tool to generate Typescript types from Go structs.

Reviewer: Thanks! I've updated my proposal based on your feedback.

-   **Expected outcomes**: Implementation of a new user setting and the infrastructure needed to support username aliases.
-   **Skills required**: React, Go
-   **Difficulty**: Hard
-   **Mentors**: Alejandro García Montoro

### Do Not Disturb improvements

Mattermost has four different values for user status: Online, Away, Do Not Disturb, and Offline.

One of the most interesting in terms of functionality (or lack thereof!) is Do Not Disturb, as it silences all the notifications while it is enabled. There is, however, still room for improvement around this feature. This project has several independent tasks, that can be implemented by the student in any order:

1. Automatic Do Not Disturb based on working hours set in Account Settings: This task would start by adding a new user setting under Account Settings to set working hours. Additionally, this setting would have a setting that would read "Automatically set my status to Do Not Disturb". When selected, the user would automatically go in and out of Do Not Disturb based on the hours defined. Users’ working hours could also be displayed in the profile popover.

> 2. Ability to notify a user in Do Not Disturb mode of important messages: Sending users will see a note above the text input area "User X is in Do Not Disturb. Notify them anyway?". Clicking "Notify them anyway" will send a special notification type that says "User X is trying to reach you". We should only allow the sending user to do this once.

Reviewer: ```suggestion
2. Ability to override a user's Do Not Disturb mode for important messages. The sender would see a note above the text input area that reads "User X is in Do Not Disturb. Notify them anyway?". Selecting "Notify them anyway" send a special notification type that says "User X is trying to reach you". We would want to limit the number of overrides sent to a user in Do Not Disturb mode.
```

> 3. Send push and desktop notification summaries when users come out of Do Not Disturb: When a user goes back online after being in Do Not Disturb mode, this feature would send a digest push and desktop notification (if enabled) like "You missed X Direct Messages and 5 channel mentions while in Do Not Disturb".

Reviewer: ```suggestion
3. Send push and desktop notification summaries when users come out of Do Not Disturb mode. When a user returns online after being in Do Not Disturb mode, this feature could send a digest push and desktop notification such as "You missed X Direct Messages and 5 channel mentions while in Do Not Disturb".
```

-   **Expected outcomes**: Implementation of the backend and frontend features for at least two out of the three tasks listed.
-   **Skills required**: React, Go
-   **Difficulty**: Medium
-   **Mentors**: Alejandro García Montoro, Eric Sethna

### Control link previews in posts

> When a message contains a URL, a preview of the linked content is shown in the post. There are situations when a plugin or an app may post several messages, each of them with a different link, which causes the channel where the messages are posted to be quickly filled with the previews. This project tries to improve this poor experience with two different tasks:

Reviewer: ```suggestion
When a message contains a URL, a preview of the linked content is shown in the post. There are situations when a plugin or an app may post several messages, each of them with a different link. This causes the channel to fill quickly with link previews. This project tries to improve this poor user experience with two different tasks:
```

1.  Modify the endpoint [`POST /posts`](https://api.mattermost.com/#tag/posts/paths/~1posts/post) in the REST API to add a property that would let the submitter control whether the posted message will contain a preview of the link in the post. This information needs to be properly communicated to the client for the preview to be enabled or disabled.
2.  Implement a new feature, probably modifying the [`PUT /posts`](https://api.mattermost.com/#tag/posts/paths/~1posts~1{post_id}/put) endpoint, to dynamically enable and disable the preview of the links in specific, existing posts.

This project is open to ideas that should be reflected in the student proposal.

> -   **Expected outcomes**: Design of the feature, modification of the endpoint to implement it and implementation of the dynamic toggling of the preview, both in the backend and the frontend.

Reviewer: ```suggestion
-   **Expected outcomes**: Design of the feature, modification of the endpoint to implement it, and implementation of the dynamic toggling of the preview, both in the backend and the frontend.
```

Author: Oh Oxford and its commas...

-   **Skills required**: React, Go
-   **Difficulty**: Easy
-   **Mentors**: Daniel Espino García, Alejandro García Montoro

### Add the concept of _Importance_ to messages

> The importance of a message varies greatly, depending on its content, the sender, the receiver or even the channel that it is being shared on. However, there is no functional difference among all these messages, and all of them are posts with the same importance for the application. This can be a problem in several situations: Think, for example, of users across different time zones or users that are on vacation; when these users are back online, they can be easily overwhelmed with the number of unread messages and notifications. There are also situations where you want to send a low priority message that does not need the immediate attention of the recipient. And others where you need a quick response or, at least, an acknowledgment.

Reviewer: ```suggestion
The importance of a message is subjective and varies greatly, depending on the content, the sender, the receiver, or even the channel where it is being shared. However, there is no functional difference between such messages, and the application regards them as having equal importance today. This can be a problem in several situations. For example, image users in different time zones or users that are on vacation -- when these users are back online, they could be easily overwhelmed with the number of unread messages and notifications. There are also situations where you want to send a low priority message that does not need the immediate attention of the recipient. And others where you need a quick response or, at least, an acknowledgment.
```

Author: Is "image users" intended here?

Reviewer: Great catch! I've updated my suggestion.

> This project is an exploratory one, to investigate how to better add the concept of importance to messages, adjusting the level of notifications on the side of the receiver, and maybe implementing a proof of concept for acknowledging a message (we in Mattermost use the ACK emoji, but it would be interesting to implement this as a standard feature): One of the original ideas here focuses on adding a small message asking the receiver if they want to acknowledge to the sender that they have read the important message.

Reviewer: ```suggestion
This project is an exploratory one to investigate how to incorporate the concept of importance to messages, adjust the level of notifications on the receiver's end, and possibly implement a proof of concept for acknowledging a message. At Mattermost, we use the ACK emoji to acknowledge messages. Perhaps this could be implemented as a standard feature. What would this workflow look like? What makes sense from both the sender's and the receiver's perspective?
```

This project is open to ideas that should be reflected in the student proposal.

> -   **Expected outcomes**: Design of the feature, proof of concept of the implementation of the concept of importance. If there is time, implementation of the feature to acknowledge important messages.

Reviewer: ```suggestion
-   **Expected outcomes**: Design of the feature, implementation proof of concept, and if there is time, implementation of the feature to acknowledge important messages.
```

-   **Skills required**: React, Go
-   **Difficulty**: Hard
-   **Mentors**: Daniel Espino García, Alejandro García Montoro

## Plugins

> Plugins are the de facto way of extending Mattermost without modifying the product itself: They are web applications on their own, whose server part directly talks to the Mattermost server, while the web part is injected in the Mattermost web application.

Reviewer: ```suggestion
Plugins are a standard way of extending Mattermost without modifying the product itself. Plugins are web applications whose server directly talks to the Mattermost server, and whose web component is injected into the Mattermost web application.
```

> Although some of them, as the Quote plugin, will require a change in the plugin infrastructure, these ideas are usually more flexible, as they are not direct modifications to the product itself, but independent pieces of software that users can decide to install and enable.

Reviewer: ```suggestion
While some of the proposed plugins will require a change within the plugin infrastructure, many of the plugin ideas are flexible, as they are not direct modifications to the product itself, but independent pieces of software that users can decide to install and enable.
```

### Spotify plugin

There are tons of integrations that are potentially useful for our users. One that seems to be asked for quite a lot is an integration with Spotify. There are several features that we would like to see in such a plugin, but the development of this idea is open to the interests of the student:

-   Share what the user is listening to in a pre-defined channel.
-   Add a new section to the profile popover that shows the current song being played.
-   Add player controls inside Mattermost: Location to be decided, but we would love to control what we are listening to without ever leaving Mattermost.
-   Collective playlist with all the songs shared to a channel.

This project is open to ideas that should be reflected in the student proposal.

-   **Expected outcomes**: A new plugin to integrate Spotify with Mattermost.
-   **Skills required**: React, Go
-   **Difficulty**: Medium
-   **Mentors**: Alejandro García Montoro

### Quote plugin

> When we are writing in a busy thread, and we want to answer a specific post, there's an unwritten rule that says that you should prepend your post with a quote of that message, formatting it as a quote with the usual Markdown syntax. This is a manual process, and as such, it is sometimes skipped. But what if we could automate it?

Reviewer: ```suggestion
When we are writing in a busy thread, and we want to answer a specific post, there's an unwritten rule that says that you should prepend your post with a quote of that message and format it as a quote with the usual Markdown syntax. This is a manual process, and as such, it gets skipped. But what if we could automate it?
```

> This project has that simple purpose: implement a new plugin that would add a new entry, `Quote`, to the post menu that would copy the corresponding post in the input box, properly formatted by prepending it with `> `. The plugin should never delete the current text in the input box, only adding this block of text.

Reviewer: ```suggestion
This project has that simple purpose: implement a new plugin that would add a new entry, `Quote`, to the post menu that would copy the corresponding post in the input box, and properly format it by prepending it with `> `. The plugin should never delete the current text in the input box, and only add the quoted block of text.
```

> However, as of today, plugins are not able to modify the input textbox, so the first and most important task in this project would be to change the plugin API to add such a feature. These architectural changes are valuable in themselves, so a project that would only deliver this would be considered successful. However, if the student finishes this task before the coding period ends, the next task would be to actually implement the quote plugin, using the new functionality as a basis for it.

Reviewer: ```suggestion
Currently, plugins aren't able to modify the input box, so the first and most important task in this project would be to change the plugin API to support such a feature. These architectural changes are valuable in themselves, so a project that would only deliver this would be considered successful. However, if the student finishes this task before the coding period ends, the next task would be to actually implement the quote plugin, using the new functionality as a basis for it.
```

-   **Expected outcomes**: The infrastructure changes for letting plugins modify the input box. If there is time, a new plugin that adds the quoting feature.
-   **Skills required**: React
-   **Difficulty**: Medium
-   **Mentors**: Daniel Espino García, Alejandro García Montoro

### Timed reminders on specific posts

> It's usual that we read an interesting message but don't have time to process it at the moment. Some of us use the "Mark as unread" feature or the TODO plugin for this, but it would be very valuable to have a feature that reminds us of that specific message in some time, generating a new notification in 30 minutes, one hour, or the next day.

Reviewer: ```suggestion
Often we read an interesting message but don't have time to process it at that moment. Some of us use the "Mark as unread" feature or the TODO plugin for this, but it would be valuable to have a feature that reminds us of a specific message after an elapsed period of time, such as after 30 minutes, one hour, or the next day.
```

> This project is focused around that specific feature, and the goal is to build a new plugin that adds a new entry to the post menu for selecting a time at which the user will be reminded of that specific message. The community has already started development on this, but it was never finished, so this is an opportunity to bring back that work and take it to the finish line, better defining its scope and usability in the process.

Reviewer: ```suggestion
This project is focused around that specific feature, and the goal is to build a new plugin that adds a new entry to the post menu that enables users to select a reminder time for the post. The community has already started development on this, but it was never finished, so this is an opportunity to bring back that work and take it to the finish line while better defining its scope and usability in the process.
```

Author: "enable users", I like that idiom!

-   **Expected outcomes**: A new plugin to add timed reminders to posts.
-   **Skills required**: React, Go
-   **Difficulty**: Easy
-   **Mentors**: Alejandro García Montoro, Eric Sethna

### Active call banner

> There are several plugins in Mattermost that integrate the product with video conference software such as Jitsi or Zoom. When a user starts a meeting, a message is automatically posted to the channel inviting other users to join. This post is then edited when the meeting ends, displaying the length of the meeting. However, the plugin post can quickly be buried in the channel if other messages are posted.

Reviewer: ```suggestion
There are several plugins in Mattermost that integrate the product with video conference software such as Jitsi or Zoom. When a user starts a meeting, a message is automatically posted to the channel inviting other users to join. This post is then edited when the meeting ends, displaying the length of the meeting. However, the plugin post can get buried in the channel as other messages are posted.
```

This project tries to overcome this problem by adding an always visible banner to the channels where there is an active meeting.

> The banner should be displayed when there's an active video call, either in Jitsi or Zoom. Clicking the banner would let the user automatically join the call, and it can display information about the meeting itself: the number of participants, current run time, whether it's in Jitsi or Zoom. This idea should be discussed with the UX team to come up with a design that could be similar to the "x new messages" that appears in an unread channel.

Reviewer: ```suggestion
The banner should be displayed when there's an active video call, either in Jitsi or Zoom. Clicking the banner would let the user automatically join the call, and it would display information about the meeting itself: the number of participants, current run time, and whether it's in Jitsi or Zoom. This idea should be discussed with the UX team to come up with a design that could be similar to the "x new messages" that appears in an unread channel.
```

-   **Expected outcomes**: Design of the feature: will it be a plugin of its own, or a modification of the Zoom and Jitsi plugin? Implementation of the banner for both cases.
-   **Skills required**: React
-   **Difficulty**: Easy
-   **Mentors**: Daniel Espino García, Alejandro García Montoro

## Apps

> Mattermost was designed from the very beginning to be extensible. Up until now, plugins were the main way (along with incoming and outgoing webhooks and slash commands) of integrating third-party services into the product, and while they have been extremely powerful and successful, we are investigating new ways of letting developers extend the functionality.

Reviewer: ```suggestion
Mattermost was designed from the very beginning to be extensible. Up until now, plugins were the main way (along with incoming and outgoing webhooks and slash commands) of integrating third-party services into the product. While they have been extremely powerful and successful, we are investigating new ways of letting developers extend Mattermost functionality.
```

> The result of that investigation is the concept of Apps: the new way to integrate external services into Mattermost.

Reviewer: ```suggestion
The result of that investigation is the concept of Apps: a new way to integrate external services into Mattermost.
```

### App Form Fields

One of the interaction points with Apps are Apps Forms. Apps Forms allow the user to introduce information to send to the App. Right now, we support some basic fields in the forms, like text, bools, or select but we would like to add more complex types and interactions, like date picker, time picker, lists, or checkbox tables.

Apps are designed with mobile support as a first class citizen, so every new form implemented has to support both the web application and the mobile apps.

> This project will be developed in close contact with the UX team to better define each individual interaction.

Reviewer: ```suggestion
This project will be developed in close collaboration with the UX team to better define each individual interaction.
```

-   **Expected outcomes**: The design and implementation of at least two new form fields. This project can be adapted to the student to implement as many form fields as desired.
-   **Skills required**: React, React Native
-   **Difficulty**: Medium
-   **Mentors**: Daniel Espino García, Alejandro García Montoro

### Apps Locations

One of the ways Mattermost can interact with the new Apps is through the App Locations: Specific locations in the UI (both in the web application and the mobile apps) that can be attached to an action handled by the App. Current locations include buttons in the channel header and the post menu.

> The goal of this project is to consider and add other locations, like entries in the main menu, additions to the channel menu or to the configuration menu.

Reviewer: ```suggestion
The goal of this project is to consider and add other locations, like entries in the main menu, additions to the channel menu, or additions to the configuration menu.
```

> One particularly interesting location to add is one to support audio rooms. There has been a lot of interest in supporting audio rooms in Mattermost, similar to the ones in Discord. There are some proofs of concept, both for Lyno and Mumble, but they are based on hacks to modify the UI of the product. This location would allow an app to host an audio only chat room in Mattermost and display that room on the channel sidebar, displaying the current status of the room.

Reviewer: ```suggestion
One particularly interesting location to consider adding is one to support audio rooms. There has been a lot of interest in supporting audio rooms in Mattermost, similar to the ones in Discord. There are some proofs of concept, both for Lyno and Mumble, but they are based on hacks to modify the UI of the product. This location would allow an app to host an audio-only chat room in Mattermost, and display that room on the channel sidebar, displaying the current status of the room.
```

> This project will be developed in close contact with the UX team to better define each individual location.

Reviewer: ```suggestion
This project will be developed in close collaboration with the UX team to better define each individual location.
```

-   **Expected outcomes**: The design and implementation of at least two new locations. This project can be adapted to the student to implement as many locations as desired.
-   **Skills required**: React, React Native
-   **Difficulty**: Medium
-   **Mentors**: Daniel Espino García, Alejandro García Montoro

## Bring your own idea

> If none of the ideas in this list speak to you, you can always propose new ones!

Reviewer: ```suggestion
If none of the ideas in this list speak to you, you can always propose your own idea!
```

Maybe you’ve been wanting to build a specific plugin for a while, you have an amazing core feature in mind that would bring the product to a whole new level, or you want to deep dive into the performance of the server to further optimize it.

> Ideas proposed by students are more than welcome, don’t hesitate to send them our way! As with all the other projects, you need to take into account that the proposals would need to fit in the overall philosophy of the project: Our product managers will take a look before we can approve them, and we can always discuss the details with the whole team if we decide to move forward with it.

Reviewer: ```suggestion
Ideas proposed by students are more than welcome, so don’t hesitate to send them our way! As with all the other projects, proposals need to fit in the overall philosophy of the project. Our Product Management team reviews each idea prior to approval, and we can always discuss the details with the whole team if we decide to move forward with it.
```
